Sources/LibrarySearchDataSource.swift

```diff
@@ -0,0 +1,105 @@
+/*****************************************************************************
+ * LibrarySearchDataSource.swift
+ * VLC for iOS
+ *****************************************************************************
+ * Copyright (c) 2019 VideoLAN. All rights reserved.
+ * $Id$
+ *
+ * Authors: Carola Nitz <caro # videolan.org>
+ *
+ * Refer to the COPYING file of the official project for license.
+ *****************************************************************************/
+
+class LibrarySearchDataSource: NSObject {
+
+    var searchData = [VLCMLObject]()
+    var model: MediaLibraryBaseModel
+
+    init(model: MediaLibraryBaseModel) {
+        self.model = model
+        super.init()
+        shouldReloadFor(searchString: "")
+    }
+
+    func objectAtIndex(index: Int) -> VLCMLObject? {
+         return index < searchData.count ? searchData[index] : nil
+    }
+
+    func shouldReloadFor(searchString: String) {
+        guard searchString != "" else {
+            searchData = model.anyfiles
+            return
+        }
+        searchData.removeAll()
+        let lowercaseSearchString = searchString.lowercased()
+        model.anyfiles.forEach {
+            if let media = $0 as? VLCMLMedia {
+                if media.contains(lowercaseSearchString) { searchData.append($0) }
+            } else if let album = $0 as? VLCMLAlbum {
+                if album.contains(lowercaseSearchString) { searchData.append($0) }
+            } else if let genre = $0 as? VLCMLGenre {
+                if genre.contains(lowercaseSearchString) { searchData.append($0) }
+            } else if let playlist = $0 as? VLCMLPlaylist {
+                if playlist.contains(lowercaseSearchString) { searchData.append($0) }
+            } else if let artist = $0 as? VLCMLArtist {
+                if artist.contains(lowercaseSearchString) { searchData.append($0) }
+            } else if let albumtrack = $0 as? VLCMLAlbumTrack {
+                if albumtrack.contains(lowercaseSearchString) { searchData.append($0) }
+            } else {
+                assertionFailure("unhandled type")
+            }
+        }
```

Comment:
In order to avoid this, since it seems that for now, we do the same thing for all the different types, what would you think of something like this:

Indeed, we could use a protocol here, like so:
```
protocol MLSearchable {
    func contains(_ searchString: String) -> Bool
}
```

Conforming the different VLCML* objects making this part quite simpler:

```
model.anyfiles.forEach {
    if let searchableFile = $0 as? MLSearchable,
        searchableFile.contains(lowercaseSearchString) {
        searchData.append($0)
     } else {
        assertionFailure("LibrarySearchDataSource: Unhandled type")
     }
}
```

Reply:
yeah sure I don't care

Reply:
yeah I believe it's actually better indeed